### PR TITLE
installation_proxy: add `install_from_bytes()`

### PR DESCRIPTION
**Pull request by abdullah-albanna**
ability to install .ipa/.ipcc using bytes objects

### why?

it's way nicer to directly install a package without dealing with paths.

someone could do something like so

```Python
import requests
from pymobiledevice3.lockdown import create_using_usbmux
from pymobiledevice3.services.installation_proxy import InstallationProxyService

file = requests.get(
    "https://raw.githubusercontent.com/Abdullah-Albanna/YemenIPCC/refs/heads/master/iPhone7Plus/iOS 15.8.2/Using Default Bundle/iPhone7Plus_iOS_15.8.2_CellularSouthLTE.ipcc",
)

lockdown = create_using_usbmux()

InstallationProxyService(lockdown).install(file.content)

```

without the need to save it to a file and then installing it and then deleting it !
too much work

this is also valid 

```Python

from pathlib import Path

file = Path("file.ipa").read_bytes()

lockdown = create_using_usbmux()

InstallationProxyService(lockdown).install(file)


```

### CODE REVIEW

**Comment by doronz88** [comment]
I much rather it be explicitly done than implicit. I'm okay with the `install_from_bytes()`, but it should a different API instead of modifying the current

**Comment by abdullah-albanna** [comment]
that makes sense, should be good now

File: pymobiledevice3/cli/apps.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def uninstall(service_provider: LockdownClient, bundle_id):
 @click.argument('package', type=click.Path(exists=True))
 def install(service_provider: LockdownServiceProvider, package: str) -> None:
     """ install given .ipa/.app/.ipcc """
-    InstallationProxyService(lockdown=service_provider).install_from_local(package)
+    InstallationProxyService(lockdown=service_provider).install(package)
```
Comment (doronz88) [inline pymobiledevice3/cli/apps.py]: there is no actual reason for this diff

File: pymobiledevice3/services/installation_proxy.py [hunk 2]
```diff
@@ -79,9 +100,9 @@ def send_cmd_for_bundle_identifier(self, bundle_identifier: str, cmd: str = 'Arc
         self.service.send_plist(cmd)
         self._watch_completion(handler, *args)
 
-    def install(self, package_path: str, options: Optional[dict] = None, handler: Callable = None, *args) -> None:
+    def install(self, package: str, options: Optional[dict] = None, handler: Callable = None, *args) -> None:
```
Comment (doronz88) [inline pymobiledevice3/services/installation_proxy.py]: restore old name

File: pymobiledevice3/services/installation_proxy.py [hunk 3]
```diff
@@ -33,6 +34,26 @@ def create_ipa_contents_from_directory(directory: str) -> bytes:
         return zip_path.read_bytes()
 
 
+def classify_zip_file(zip_bytes: bytes) -> Literal['ipa', 'ipcc']:
```
Comment (doronz88) [inline pymobiledevice3/services/installation_proxy.py]: use python's Enum instead

File: pymobiledevice3/services/installation_proxy.py [hunk 4]
```diff
@@ -33,6 +34,26 @@ def create_ipa_contents_from_directory(directory: str) -> bytes:
...
+    """ checks the zipped bytes if it's a .ipcc or .ipa """
+    try:
+        with ZipFile(BytesIO(zip_bytes), 'r') as zip_file:
+            # sometimes packages at first index don't have enough infos to check
+            dirs = zip_file.namelist()[1].split('/')
+
+            if dirs[0] != 'Payload':
+                raise AppInstallError('package does not have a payload')
```
Comment (doronz88) [inline pymobiledevice3/services/installation_proxy.py:57]:
```suggestion
                raise InvalidArgumentError('package does not have a payload')
```

File: pymobiledevice3/services/installation_proxy.py [hunk 5]
```diff
@@ -33,6 +34,26 @@ def create_ipa_contents_from_directory(directory: str) -> bytes:
...
+            if dirs[1].endswith('.app'):
+                return 'ipa'
+            elif dirs[1].endswith('.bundle'):
+                return 'ipcc'
+            else:
+                raise AppInstallError('package does not have the appropriate folders structure')
+
+    except BadZipFile:
+        raise AppInstallError('invalid bytes package')
```
Comment (doronz88) [inline pymobiledevice3/services/installation_proxy.py:66]:
```suggestion
        raise InvalidArgumentError('invalid bytes package')
```

File: pymobiledevice3/services/installation_proxy.py [hunk 6]
```diff
@@ -96,18 +117,38 @@ def uninstall(self, bundle_identifier: str, options: Optional[dict] = None, hand
...
         """ upload given ipa/ipcc onto device and install it """
-
         ipcc_mode = package_path.suffix == '.ipcc'
 
         if options is None:
             options = {}
 
         if ipcc_mode:
             options['PackageType'] = 'CarrierBundle'
+
```
Comment (doronz88) [inline pymobiledevice3/services/installation_proxy.py]: remove this diff

File: pymobiledevice3/services/installation_proxy.py [hunk 7]
```diff
@@ -124,20 +165,35 @@ def install_from_local(self, package_path: Path, cmd: str = 'Install', options:
...
+            'PackagePath': (
+                TEMP_REMOTE_IPCC_FOLDER if ipcc_mode
+                else TEMP_REMOTE_IPA_FILE
+            )
+        })
 
         self._watch_completion(handler, ipcc_mode, args)
 
-    def upload_ipcc_as_folder(self, file: Path, afc_client: AfcService) -> None:
+    def upload_ipcc_as_folder(self, file: Union[Path, bytes], afc_client: AfcService) -> None:
```
Comment (doronz88) [inline pymobiledevice3/services/installation_proxy.py]: restore this function behavior to the old one. the bytes objects can already be convered to file in `install_from_bytes()`

**Review by doronz88 — CHANGES_REQUESTED** [review]

File: pymobiledevice3/cli/apps.py (hunk 1 above)
Comment (abdullah-albanna) [reply pymobiledevice3/cli/apps.py]: what's the point of .install()?

File: pymobiledevice3/services/installation_proxy.py (hunk 4 above)
Comment (abdullah-albanna) [reply pymobiledevice3/services/installation_proxy.py:57]: where is that exception?

File: pymobiledevice3/services/installation_proxy.py (hunk 6 above)
Comment (abdullah-albanna) [reply pymobiledevice3/services/installation_proxy.py]: I don't understand, remove what?

File: pymobiledevice3/services/installation_proxy.py (hunk 7 above)
Comment (abdullah-albanna) [reply pymobiledevice3/services/installation_proxy.py]: what do you mean?,  .ipcc  must be uploaded as a folder for both of the bytes and the path

correct me if that's not what you mean

**Comment by doronz88** [comment]
Just making sure you saw I left some comments still

**Comment by abdullah-albanna** [comment]
sorry, I don't get what you mean

can you tell me what to edit please, I understood some of it, but others are kind of confusing

**Comment by doronz88** [comment]
There are still unresolved issues. See the existing comments

**Comment by abdullah-albanna** [comment]
Yeah but could you take a look at my replies please 

The unresolved ones are what confuses me, I have already put a reply to them

**Comment by doronz88** [comment]
image
I don't see any reply to any comment

**Comment by abdullah-albanna** [comment]
my bad, I did not commit them

File: pymobiledevice3/services/installation_proxy.py (hunk 6 above)
Comment (doronz88) [reply pymobiledevice3/services/installation_proxy.py]: this extra line what was added in the PR

File: pymobiledevice3/services/installation_proxy.py (hunk 7 above)
Comment (doronz88) [reply pymobiledevice3/services/installation_proxy.py]: i don't want an api function acting differently depending in its supplied arguments. if it performs another logic it should be in its own api. the expected `file` argument can remain `Path`

**Review by doronz88 — CHANGES_REQUESTED** [review]

**Comment by abdullah-albanna** [comment]
I like how it gets crazy over an unused import

**Review by doronz88 — APPROVED** [review]

**Comment by doronz88** [comment]
Thanks for your contribution!

**Pull request merged**